### PR TITLE
Support mzQC validation

### PR DESCRIPTION
**Pull request by cbielow**
via use of `validateFrom..()` functions.
Test and Vignette are also added.

Also:
fix empty pkgdown Changelog site, due to misformatted version number (see https://github.com/r-lib/pkgdown/issues/456), i.e. do not use **vX.X.X** in `NEWS.md` file, but something like **version X.X.X** instead

### CODE REVIEW

**Review by biodavidjm — COMMENTED** [review]
Everything looks good. 

However, I have a suggestion. What about having a unique function `validate_mzqc` which internally handles the different inputs, i.e., from either file, object, or string?

**Comment by cbielow** [comment]
Thanks for the review :)

> Everything looks good.
> 
> However, I have a suggestion. What about having a unique function `validate_mzqc` which internally handles the different inputs, i.e., from either file, object, or string?

This is what I tried to do, by having an internal function which creates the validator using our mzQC schema. And the three validation functions call themselves, when possible.
Is there a way to make this even more compact?  Suggestions welcome :)

File: R/validator.R [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+#'
+#' @param mzqc_root An mzQC root object
+#' @param verbose Show extra information if validation fails
+#' @return TRUE/FALSE if validation was successful/failed
+#'
+#' @export
+#'
+validateFromObj = function(mzqc_root, verbose = TRUE)
+{
+  if (!("MzQCmzQC" %in% class(mzqc_root))) stop("Argument is not a class of 'MzQCmzQC'.")
```
Comment (mwalzer) [inline R/validator.R:82]: Sounds reasonable. Maybe an additional check if toJSON was successful (i.e. valid json)?

File: R/validator.R [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+#'
+#' @param filepath A path to a file (e.g. "c:/my.mzQC", or "test.mzQC")
+#' @param verbose Show extra information if validation fails
+#' @return TRUE/FALSE if validation was successful/failed
+#'
+#' @export
+#'
+validateFromFile = function(filepath, verbose = TRUE)
+{
+  data = readLines(filepath)
```
Comment (mwalzer) [inline R/validator.R:34]: check and report if reading was successful would be useful.

**Review by mwalzer — CHANGES_REQUESTED** [review]
Check jsonlite::toJSON/read from file was working would be useful.
Maybe mention somewhere it's syntax validation. (Unless I overlooked the semantic check, but I think you mentioned as much in a telco.)

File: R/validator.R (hunk 1 above)
Comment (cbielow) [reply R/validator.R:82]: This is done 'automatically' during `::toJSON`. Users will see an error message and a backtrace.

File: R/validator.R (hunk 2 above)
Comment (cbielow) [reply R/validator.R:34]: there is a default error message if `readlines()` fails, but its not very good:
```
> rmzqc::validateFromFile("doesnotExist.mzQC")
Error in file(con, "r") : cannot open the connection
In addition: Warning message:
In file(con, "r") :
 
 Error in file(con, "r") : cannot open the connection
```

I've added a check which now says
```
> validateFromFile("doesnotExist.mzQC")
Error in validateFromFile("doesnotExist.mzQC") : 
  The file 'doesnotExist.mzQC' does not exist (or is not readable).
```

File: R/validator.R (hunk 1 above)
Comment (cbielow) [reply R/validator.R:82]: However, this reminds me that one could check if the JSON itself when reading from a file/string is valid.
Done.

Before (not very informative)
```
res = validateFromString(strings)
Error in context_eval(join(src), private$context, serialize, await) : 
SyntaxError: Invalid or unexpected token
```

After:
```
>   res = validateFromString(strings)
Warning in validateFromString(strings) :
  925lexical error: invalid character inside string.
           software for QC report genera             }           ]    
                     (right here) ------^

Error in validateFromString(strings) : 
  The JSON string is not valid JSON.
  ```

Comment (mwalzer) [reply R/validator.R:91]: :+1

File: R/validator.R (hunk 2 above)
Comment (mwalzer) [reply R/validator.R:37]: 👍

**Review by mwalzer — APPROVED** [review]
Good improvements

**Pull request merged**